### PR TITLE
Join sort optims

### PR DESCRIPTION
**Pull request by steadtler**
Added a new algo to perform a quicksort in multithread.

On fmaillet's test case of a left join on 41M+ rows (on one side, less on the other), with this PR we save:

About 60 seconds by sorting in parallel in join_dataset
About 173 seconds (3 minutes!) by sorting in parallel in QueryBasic
About 20 seconds by not sorting in TabularDataset::getrownames

In total we cut the time necessary to perform this large outer join by more than half. There may be a larger refactoring to make it much faster but those were the low-hanging apples.

### CODE REVIEW

File: plugins/tabular_dataset.cc [hunk 1]
```diff
@@ -257,13 +257,6 @@ struct TabularDataset::TabularDataStore: public ColumnIndex, public MatrixView {
             result.insert(result.end(), c.rowNames.begin(), c.rowNames.end());
         }
 
-        std::sort(result.begin(), result.end(),
```
Comment (jeremybarnes) [inline plugins/tabular_dataset.cc]: This one makes me nervous.

**Comment by jeremybarnes** [comment]
It all looks good, but I wonder about not sorting the row names.  If we don't here, we shouldn't anywhere and we should then not expect anything to be sorted in the rest of the code.

**Comment by steadtler** [comment]
Its my next story to remove it everywhere else. I can wait until that story is done if you want.

**Comment by jeremybarnes** [comment]
I'm OK to merge it, so long as you document in dataset.h what the requirement on the method is, and also in the places where it's called.  You should also document the ordering requirement for the RowStream.

**Comment by steadtler** [comment]
@jeremybarnes can you review the updated code documentation?

**Comment by guyd** [comment]
I have a related question, do all datasets have a natural row ordering (e.g. the order in which recordRow[s] was called)?  If so, what is that natural order and do `MatrixView::getRowNames` and `ColumnIndex::getRowNames` return rows in that natural order.

**Comment by steadtler** [comment]
@guy 

Yes, they should all have a natural ordering of sorts. getrownames can return the rows in any order as long as successive calls will return them in the same order.

**Comment by guyd** [comment]
So the order is dataset-dependent but consistent for a given dataset.

File: core/dataset.h [hunk 2]
```diff
@@ -60,6 +60,18 @@ typedef EntityType<Dataset> DatasetType;
 struct MatrixView {
     virtual ~MatrixView();
 
+    /*
```
Comment (jeremybarnes) [inline core/dataset.h]: /** for Doxygen

File: core/dataset.h [hunk 3]
```diff
@@ -420,6 +436,8 @@ struct Dataset: public MldbEntity {
         Must return the *exact* set of rows or a stream that will do the same
         because the where expression will not be evaluated outside of this method
         if this method is called.
+
+        Ordering can be arbitrary but need to be deterministic
```
Comment (jeremybarnes) [inline core/dataset.h]: needs to be, missing fullstop (minor)

Comment (jeremybarnes) [inline core/dataset.h]: Also, we should say that they must not be duplicated.

**Comment by jeremybarnes** [comment]
I had some minor comments on the doc.  My +1 still stands once you've lokoed at them.

In the future, we should look at having each dataset be able to expose its natural row order.  This can lead to all kinds of optimizations, especially for joining of two datasets that have the same natural order.

**Pull request merged**